clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java

```diff
@@ -93,6 +83,83 @@ public Void operate(List priorOutputs) {
     return null;
   }
 
+  @VisibleForTesting
+  public UpsertEntityTagsDescription setIdAndIdPattern(UpsertEntityTagsDescription description) {
+    EntityTags.EntityRef entityRef = description.getEntityRef();
+    String entityRefAccount = (String) entityRef.attributes().get("account");
+    String entityRefAccountId = (String) entityRef.attributes().get("accountId");
+    if (entityRefAccount != null && entityRefAccountId == null) {
+      // add `accountId` if available (and not already specified)
+      AccountCredentials accountCredentials = accountCredentialsProvider.getCredentials(entityRefAccount);
+      if (accountCredentials != null) {
+        entityRefAccountId = accountCredentials.getAccountId();
+        entityRef.attributes().put("accountId", entityRefAccountId);
+      }
+    }
+
+    EntityRefIdBuilder.EntityRefId entityRefId = EntityRefIdBuilder.buildId(
+      entityRef.getCloudProvider(),
+      entityRef.getEntityType(),
+      entityRef.getEntityId(),
+      Optional.ofNullable(entityRefAccountId).orElse(entityRefAccount),
+      (String) entityRef.attributes().get("region")
+    );
+    description.setId(entityRefId.id);
+    description.setIdPattern(entityRefId.idPattern);
+    return entityTagsDescription;
+  }
+
+  @VisibleForTesting
+  public Void mergeExistingTagsAndMetadata(EntityTags currentTags, EntityTags updatedTags) {
+    String user = AuthenticatedRequest.getSpinnakerUser().orElse("unknown");
+    Long now = System.currentTimeMillis();
+
+    if (currentTags == null) {
+      addTagMetadata(updatedTags);
+    } else {
+      updatedTags.setTagsMetadata(currentTags.getTagsMetadata() == null ?
+        new HashMap<>() :
+        currentTags.getTagsMetadata());
+
+      updatedTags.getTags().forEach((key, tag) -> {
+        EntityTags.EntityTagMetadata metadata = currentTags.getTagsMetadata().get(key);
+        if (metadata == null) {
+          metadata = new EntityTags.EntityTagMetadata();
+          metadata.setCreated(now);
+          metadata.setCreatedBy(user);
+          updatedTags.getTagsMetadata().put(key, metadata);
+        }
+        metadata.setLastModified(now);
+        metadata.setLastModifiedBy(user);
+      });
+      currentTags.getTags().forEach((key, tag) -> {
+        if (!updatedTags.getTags().containsKey(key)) {
+          updatedTags.getTags().put(key, tag);
+          updatedTags.getTagsMetadata().put(key, currentTags.getTagsMetadata().get(key));
+        }
+      });
+    }
+    return null;
+  }
+
+  private Void addMetadataForTag(EntityTags entityTags, String key) {
```

Comment:
`private static EntityTags.EntityTagMetadata tagMetadata()`

and then on the caller side ...

`entityTags.getTagsMetadata().put(key, tagMetadata())`